### ISSUE TITLE
Support for RFM69

### ISSUE THREAD

**Issue by pc-coholic** [OpenHR20/OpenHR20#27]
I am starting to RFMifying my HR20s and was running into the issue, that it has become more and more difficult to acquire RFM12-modules directly in Asia, as the manufacturer has ceased production of them...

The new go-to RFM seems to be the RFM69 - the cw-variant is even the same size as the RFM12.

Unfortunately, they are not a drop-in-replacement - as far as I was able to see, they need at least a different initialization-sequence. But properly initialized, they can even co-exists and communicate with RFM12-devices.

Therefore I wanted to inquire if there are any plans on adding support for RFM69-radios through a compile-switch. As I do not have a lot experience with wireless radios, I don't think that I'd be able to port the necessary changes myself.

However, if someone else would like to do it, I am more than happy to provide a RFM69-module for appropriate testing.

**Comment by rp42** [comment]
I bought RFM69 by accident and I thought about having a go at adding
support. But realised I didn't really have the time. I bought RFM12
instead, but have yet to get them working...

See here about running both modules in the same network:
http://jeelabs.org/2015/05/27/rfm69-on-atmega/

Rhys

**Comment by KickerTom** [comment]
Hi,

I started adding RFM69 support a while back, but never finished that due to lack of time. I still plan to get back to it, but can't promise any dates, sorry.
If you need some RFM12Bs, I still have a box of them here :-) (mainly 433MHz).

Tomas

**Comment by diasarmando** [comment]
Hi

There is update about the rfm69?

thanks

**Comment by KickerTom** [comment]
Hi,
thanks for your interest. I still have rfm69 support on my list, I even started implementing it, but due to lack of time, it will probably not materialize in any near future, sorry. But PRs are always welcome :-).

**Comment by dewi-ny-je** [comment]
Would it be difficult to simply output the data for the wireless transmission via SPI, without caring about the type of module, so that the daughter Arduino board including the wireless module of choice can take care of the transmission? at this point it could even be a ESP8266, it's up to the user to connect something.

**Comment by diasarmando** [comment]
The ESP8266 is a bad solutions because has a bad efficiency compared with the other radio solutions we talked before, the ESP8266 have 80mA as average consumption.

**Comment by dewi-ny-je** [comment]
It was just an example, attaching a rfm69 of course would be better. Still, having the option to send data to SPI would allow the use of any module, present or future.

**Comment by KickerTom** [comment]
Well, everything is possible, and sending only data would not be that difficult. However you still need to add some control information (do you want to send data or receive data? Do you want to get some information about signal strength? etc.), so you need to come up with a protocol. Considering that your suggestion also adds more hardware (currently HR20 talks to RFM12 directly via SPI, and you recommend adding arduino board in between), and that you just shift the RFM programming from HR20 to arduino, and add a new protocol in between, I fail to see any advantage.
Maybe just that you can use existing RFM69 libraries for arduino, but the glue programming in between is probably more work than adapting the arduino libraries for RFM20.

**Comment by diasarmando** [comment]
I sugest this:
- Modify the firmware to support the RFM69 because the RFM12 is not produced anymore
- Use a mqtt Gateway like this one: http://tinkerman.cat/rfm69-wifi-gateway/
- Do a very small PCB adapter to connect the RFM69 in to the HR20/25 programming header. we could also do an external small shell in order to put the radio module as an addon to the HR20/25.

I can start doing the small PCB and the small sheell... :)

**Comment by diasarmando** [comment]
By the way, using the mqtt gateway, you will can use every IOT platmort you want, like home assistant, we just need to use, adapt or even create a module for it.

**Comment by MalteP** [comment]
RFM12B is still in production and will not be discontinued in the coming years according to the HopeRF website: https://www.hoperf.com/modules/rf_transceiver/RFM12BW.html

I also spent some days debugging, but my RFM12 setup is running more than a year now.
Please have a look at this post: https://groups.google.com/forum/#!topic/openhr20-development/j4YzIwSXR3Y

The MQTT stuff you suggested should be done on the master / host side because the time slot based protocol implementation saves a lot of power. My HR20s are still running on the first set of alkaline batteries.

**Comment by diasarmando** [comment]
@MalteP It was my missunderstandin because its not easy to find it available at resonable prices... 

REgardint the MQTT stuff, of course Im refereiing to do it outside of HR20s, needs to be done on the gateway, the gateway/brigde that will post the mqtt queries on the mqtt server.

**Comment by G-A-S** [comment]
Hi @KickerTom, did you had in meanwhile some time to review the RFM69 support? Thx!

**Comment by c-MM** [comment]
Hi,
I think there's no real advantage to add RFM69 support.

I'm currently building a new master system based on RFM95 (the ones that than can do FSK and LORA). 

The FSK-mode of the RFM95 seems to be pretty similar to the RFM69 but it actually seems less reliable than the RFM12b. I run 18 OpenHR20 in my house since over 12 years with RFM12b internally wired. Yet I have not succeeded to find a config for the RFM95 that offers the same reliability as the old RFM12b master. It works, but it looses a lot more packets than the old master. On the other side it was also a very long way to get the RFM12bs to work reliably.

The RFM95 is directly attached to the GPIOs of a raspberry pi and I'm rewriting the php-code and master c-code in lua. I have xtea and cmac crypto successfully implemented in lua, but the code for transmission of packets is still missing (i only receive the packets from the thermostats and the old master and decode them). This is mainly because I want to move the master to my openwrt router (which also has SPI GPIOs available) and there is already lua on it installed since openwrt uses lua for scripting. I'll probably add mqtt to control the HR20s from HomeAssistant and drop the old web interface. Not sure yet.

When the new lua master code is working I will probably try to modify a HR20 with RFM95 module (if I can find an additionally accessible pin on the AVR, since the RFM95 (and also the RFM69) seems to need one more GPIO to be connected to the AVR) and try the LORA modulation and see if it's more reliable than the RFM12b. If not, I'll stick to the RFM12bs and won't bother to resolder all my HR20s.

Just my 2 cents. your mileage might vary.

**Comment by dewi-ny-je** [comment]
I wonder whether so much work even makes sense since these devices are nowadays difficult to source for new installations, and if yours work fine... there are bridges which allow to connect hr20s to mqtt which might be even easier than fiddling with new RFM modules.

Personally I never modified mines since my new house is so well insulated that I simply set them to manual value depending on the room and they stay that way the whole season.
And if I needed more control... There are ZigBee ones starting at 16 euro.

**Comment by c-MM** [comment]
You are right. I'm testing the RFM95 modules just out of curiosity. Even if the LORA modulation proves to be much better, I still don't know if I'll change all my HR20s.

But I like the HR20s since I have them and since I love the 2.5 years of runtime I get out of two Eneloop NiMhs.

I also like that I was able to tune them to match my rooms and provide very good regulation. But I invested a lot of time in tuning the values. Since I will change my heating system from gas to a heat pump next year, I decided to modernize the rest of my heating control too.

I have fun doing that, that's why it makes sense for me. :-)
